cmd/kubeadm/app/util/patches/patches.go

```diff
@@ -0,0 +1,338 @@
+			patchBytes := []byte(patch)
+
+			// Patch based on the patch type.
+			switch patchSet.patchType {
+
+			// JSON patch.
+			case types.JSONPatchType:
+				var patchObj jsonpatch.Patch
+				patchObj, err = jsonpatch.DecodePatch(patchBytes)
+				if err == nil {
+					patchedData, err = patchObj.Apply(patchedData)
+				}
+
+			// Merge patch.
+			case types.MergePatchType:
+				patchedData, err = jsonpatch.MergePatch(patchedData, patchBytes)
+
+			// Strategic merge patch.
+			case types.StrategicMergePatchType:
+				patchedData, err = strategicpatch.StrategicMergePatch(
+					patchedData,
+					patchBytes,
+					patchTarget.StrategicMergePatchObject,
+				)
+			}
+
+			if err != nil {
+				return errors.Wrapf(err, "could not apply the following patch of type %q to target %q:\n%s\n",
+					patchSet.patchType,
+					patchTarget.Name,
+					patch)
+			}
+			fmt.Fprintf(pm.output, "[patches] Applied patch of type %q to target %q\n", patchSet.patchType, patchTarget.Name)
+		}
+
+		// Update the data for this patch target.
+		patchTarget.Data = patchedData
+	}
+
+	return nil
+}
+
+// getTargetNameFromFilename accepts a file name and returns a known target name string,
+// or an error if the target name is unknown.
+func getTargetNameFromFilename(fileName string, knownTargets []string) (string, error) {
+	var targetName string
+
+	for _, target := range knownTargets {
+		if strings.HasPrefix(fileName, target) {
+			targetName = target
+			break
+		}
+	}
+
+	if len(targetName) == 0 {
+		return "", errors.Errorf("received file name %q, but target must be one of %v", knownTargets, fileName)
+	}
+	return targetName, nil
+}
+
+// getPatchTypeFromFilename accepts a file name and returns the patch type encoded in it.
+// For example, "etcd+merge.json" would return "merge". Returns an error on a unknown patch types.
+func getPatchTypeFromFilename(fileName string) (types.PatchType, error) {
+	idxDot := strings.LastIndex(fileName, ".")
+	if idxDot == -1 {
+		// Shift the dot index into the length of the string.
+		idxDot = len(fileName)
+	}
+
+	idxPlus := strings.Index(fileName, "+")
+	if idxPlus == -1 || idxPlus > idxDot {
+		// If the + character is missing or if its index is after the dot,
+		// just shift it into the dot index.
+		idxPlus = idxDot
+	} else {
+		// Increment the plus index to discard the + character itself.
+		idxPlus++
+	}
+
+	patchType := fileName[idxPlus:idxDot]
+	pt, ok := patchTypes[patchType]
+	if !ok {
+		return "", errors.Errorf("unknown patch type %q from file name %q", patchType, fileName)
+	}
+	return pt, nil
+}
+
+// createPatchSet creates a patchSet object, by splitting the given "data" by "\n---".
+func createPatchSet(targetName string, patchType types.PatchType, data string) (*patchSet, error) {
+	var patches []string
+
+	if len(data) > 0 {
+		// Split the patches and convert them to JSON.
+		// Data that is already JSON will not cause an error.
+		patches = strings.Split(data, "\n---")
```

Comment:
should we use `YAMLReader` from "k8s.io/apimachinery/pkg/util/yaml" for this?

Reply:
+1 to use a `YAMLReader` here. `SplitYAMLDocuments` also uses it so we won't be pulling a new package here.

Reply:
i anticipated this question.

so i compared https://github.com/kubernetes/apimachinery/blob/master/pkg/util/yaml/decoder.go#L146 this to a plain `strings.Split(data, "\n---")` and honestly i couldn't find a difference or a case where a certain input would work for `splitYAMLDocument()` but not for `strings.Split(data, "\n---")`.

so my question here is why do we need `YAMLReader`?
also note that the input at that point can be technically both YAML/JSON.

Reply:
my comment above applies the same way for our SplitYAMLDocuments and apimachinery's splitYAMLDocument.

i don't not see (couldn't find) the benefits over `strings.Split(data, "\n---")`.

Reply:
FYI, in clusterctl we started with strings.Split(data, "\n---") and then reverted to YAMLReader after reports of error in some corner cases (I remember the cert-manager YAML being one of them).

Reply:
It would probably be best to have bug-2-bug compatibility with the rest of the k8s ecosystem here (and kubeadm itself, since `SplitYAMLDocuments` uses a `YAMLReader`).

Reply:
i will switch to YAMLReader even if we are going to feed JSON to it.

@fabriziopandini 

> then reverted to YAMLReader after reports of error in some corner cases (I remember the cert-manager YAML being one of them).

do you have the link to these reports?